src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/YFlowUpdateService.java

```diff
@@ -0,0 +1,387 @@
+                    }
+
+                    subFlowBuilder.description(subFlowPartialUpdate.getDescription());
+
+                    subFlow = subFlowBuilder.build();
+                }
+                subFlows.add(subFlow);
+            }
+            target.setSubFlows(subFlows);
+        }
+
+        target.setType(Type.UPDATE);
+
+        handleRequest(key, commandContext, target);
+    }
+
+    /**
+     * Handles async response from worker.
+     *
+     * @param key command identifier.
+     */
+    public void handleAsyncResponse(String key, SpeakerFlowSegmentResponse flowResponse) throws UnknownKeyException {
+        log.debug("Received flow command response {}", flowResponse);
+        YFlowUpdateFsm fsm = getFsmByKey(key)
+                .orElseThrow(() -> new UnknownKeyException(key));
+
+        String flowId = flowResponse.getMetadata().getFlowId();
+        if (fsm.getUpdatingSubFlows().contains(flowId)) {
+            flowUpdateService.handleAsyncResponseByFlowId(flowId, flowResponse);
+        } else {
+            YFlowUpdateContext context = YFlowUpdateContext.builder()
+                    .speakerResponse(flowResponse)
+                    .build();
+            if (flowResponse instanceof FlowErrorResponse) {
+                fsmExecutor.fire(fsm, Event.ERROR_RECEIVED, context);
+            } else {
+                fsmExecutor.fire(fsm, Event.RESPONSE_RECEIVED, context);
+            }
+        }
+
+        // After handling an event by FlowUpdate service, we should propagate execution to the FSM.
+        fsmExecutor.fire(fsm, Event.NEXT);
+
+        removeIfFinished(fsm, key);
+    }
+
+    /**
+     * Handles timeout case.
+     *
+     * @param key command identifier.
+     */
+    public void handleTimeout(String key) throws UnknownKeyException {
+        log.debug("Handling timeout for {}", key);
+        YFlowUpdateFsm fsm = getFsmByKey(key)
+                .orElseThrow(() -> new UnknownKeyException(key));
+
+        // Propagate timeout event to all sub-flow processing FSMs.
+        fsm.getUpdatingSubFlows().forEach(flowId -> {
+            try {
+                flowUpdateService.handleTimeoutByFlowId(flowId);
+            } catch (UnknownKeyException e) {
+                log.error("Failed to handle a timeout event by FlowCreateService for {}.", flowId);
+            }
+        });
+
+        fsmExecutor.fire(fsm, Event.TIMEOUT);
+
+        removeIfFinished(fsm, key);
+    }
+
+    private void removeIfFinished(YFlowUpdateFsm fsm, String key) {
+        if (fsm.isTerminated()) {
+            log.debug("FSM with key {} is finished with state {}", key, fsm.getCurrentState());
+            unregisterFsm(key);
+
+            carrier.cancelTimeoutCallback(key);
+            if (!isActive() && !hasAnyRegisteredFsm()) {
+                carrier.sendInactive();
+            }
+        }
+    }
+
+    private String generateFlowId(String prefix) {
```

Comment:
Should/can it be shared with y-flow create "service"?

Reply:
Agree, we need to move it in some shared class.

Reply:
Fixed.